src/ext/php7/engine_hooks.c

```diff
@@ -479,80 +448,362 @@ static int _dd_opcode_default_dispatch(zend_execute_data *execute_data) {
+    zend_function *current_fbc = EX(call)->func;
+    DDTRACE_G(original_context).fbc = current_fbc;
+    zend_function *previous_calling_fbc = DDTRACE_G(original_context).calling_fbc;
 
-        zend_object *previous_this = DDTRACE_G(original_context).this;
-        DDTRACE_G(original_context).this = this ? Z_OBJ_P(this) : NULL;
-        zend_class_entry *previous_calling_ce = DDTRACE_G(original_context).calling_ce;
+    DDTRACE_G(original_context).calling_fbc = current_fbc->common.scope ? current_fbc : execute_data->func;
 
-        if (DDTRACE_G(original_context).this) {
-            GC_ADDREF(DDTRACE_G(original_context).this);
-        }
-        DDTRACE_G(original_context).calling_ce = Z_OBJ(execute_data->This) ? Z_OBJ(execute_data->This)->ce : NULL;
+    zval *this = _dd_this(EX(call));
 
-        _dd_wrap_and_run(execute_data, dispatch);
-        if (DDTRACE_G(original_context).this) {
-            GC_DELREF(DDTRACE_G(original_context).this);
-        }
+    zend_object *previous_this = DDTRACE_G(original_context).this;
+    DDTRACE_G(original_context).this = this ? Z_OBJ_P(this) : NULL;
+    zend_class_entry *previous_calling_ce = DDTRACE_G(original_context).calling_ce;
 
-        // Restore original context
-        DDTRACE_G(original_context).calling_ce = previous_calling_ce;
-        DDTRACE_G(original_context).this = previous_this;
-        DDTRACE_G(original_context).calling_fbc = previous_calling_fbc;
-        DDTRACE_G(original_context).fbc = previous_fbc;
+    if (DDTRACE_G(original_context).this) {
+        GC_ADDREF(DDTRACE_G(original_context).this);
+    }
+    DDTRACE_G(original_context).calling_ce = Z_OBJ(execute_data->This) ? Z_OBJ(execute_data->This)->ce : NULL;
 
-        _dd_update_opcode_leave(execute_data);
+    _dd_wrap_and_run(execute_data, dispatch);
+    if (DDTRACE_G(original_context).this) {
+        GC_DELREF(DDTRACE_G(original_context).this);
     }
 
+    // Restore original context
+    DDTRACE_G(original_context).calling_ce = previous_calling_ce;
+    DDTRACE_G(original_context).this = previous_this;
+    DDTRACE_G(original_context).calling_fbc = previous_calling_fbc;
+    DDTRACE_G(original_context).fbc = previous_fbc;
+
+    _dd_update_opcode_leave(execute_data);
+
     dispatch->busy = 0;
-    ddtrace_class_lookup_release(dispatch);
+    ddtrace_dispatch_release(dispatch);
 
     EX(opline)++;
 
     return ZEND_USER_OPCODE_LEAVE;
 }
 
+static int _dd_do_icall_handler(zend_execute_data *execute_data) {
+    // Assume that a neighboring extension could have incremented the opline before forwarding the handler
+    if (ZEND_DO_ICALL != EX(opline)->opcode) {
+        return _prev_icall_handler ? _prev_icall_handler(execute_data) : ZEND_USER_OPCODE_DISPATCH;
+    }
+    ddtrace_dispatch_t *dispatch = NULL;
+    if (!EX(call)->func || !_dd_should_trace_call(EX(call), EX(call)->func, &dispatch)) {
+        return _prev_icall_handler ? _prev_icall_handler(execute_data) : ZEND_USER_OPCODE_DISPATCH;
+    }
```

Comment:
I think the following change is equivalent. If so, we can apply the same technique to all of the call handlers.
```suggestion
    ddtrace_dispatch_t *dispatch = NULL;
    if (ZEND_DO_ICALL != EX(opline)->opcode || !EX(call)->func || !_dd_should_trace_call(EX(call), EX(call)->func, &dispatch)) {
        return _prev_icall_handler ? _prev_icall_handler(execute_data) : ZEND_USER_OPCODE_DISPATCH;
    }
```

Reply:
Excellent suggestion! I've refactored them with your suggestion. I also added the opcode check that we discussed yesterday to the other handlers so that we have the same safety net across the board.